Compiler complains/warns about the use of a substitution variable for deep sleep duration

### The problem

In Yaml, I have:
```yaml
substitutions:
  dozetime: "60min"
  criticalsleep: "24h"
```
....
```yaml
      - if: 
          condition: { lambda: 'return esp_sleep_get_wakeup_cause() != 7;' }
          then:
            - if:
                condition: 
                  switch.is_off: service_mode_switch
                then:            
                  - lambda: 'id(sleeping_state) = true;'
                  - deep_sleep.enter: 
                      id: sleep_mode
                      sleep_duration: ${dozetime}
```

When I try to comple using esphome builder, It doesn't like the use of dozetime here although it compiles successfully anyway


>  In file included from src/esphome/components/sensor/filter.h:9,
>                  from src/esphome/components/sensor/sensor.h:8,
>                  from src/esphome/components/adc/adc_sensor.h:3,
>                  from src/esphome.h:3,
>                  from src/main.cpp:3:
> src/esphome/components/deep_sleep/deep_sleep_component.h: In instantiation of 'void esphome::deep_sleep::EnterDeepSleepAction<Ts>::set_sleep_duration(V) [with V = setup()::<lambda()>; Ts = {}]':
> /config/esphome/fall-detector.yaml:967:59:   required from here
>   967 |                 sleep_duration: ${criticalsleep}
>       |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~          ^
>   968 |           else:
>       |           ~~~~~                                            
>   969 |             - if:
>       |                                                            
> src/esphome/core/automation.h:108:64: warning: 'esphome::TemplatableFn<T, X>::TemplatableFn(F) requires !(convertible_to<F, T (*)('type_pack_expansion' not supported by simple_type_specifier'type_pack_expansion' not supported by direct_abstract_declarator)>) && (invocable<F, X ...>) && (convertible_to<typename std::invoke_result<F, X ...>::type, T>) && (is_empty_v<F>) && (default_initializable<F>) [with F = setup()::<lambda()>; T = long unsigned int; X = {}]' is deprecated: Lambda return type does not match TemplatableFn<T> — use the correct type in codegen [-Wdeprecated-declarations]
>   108 |   template<typename V> void set_##name(V name) { this->name##_ = name; }
> src/esphome/core/automation.h:110:39: note: in expansion of macro 'TEMPLATABLE_VALUE_'
>   110 | #define TEMPLATABLE_VALUE(type, name) TEMPLATABLE_VALUE_(type, name)
>       |                                       ^~~~~~~~~~~~~~~~~~
> src/esphome/components/deep_sleep/deep_sleep_component.h:167:3: note: in expansion of macro 'TEMPLATABLE_VALUE'
>   167 |   TEMPLATABLE_VALUE(uint32_t, sleep_duration);
>       |   ^~~~~~~~~~~~~~~~~
> src/esphome/core/automation.h:54:33: note: declared here
>    54 |                    "codegen")]] TemplatableFn(F) requires(!std::convertible_to<F, T (*)(X...)>) &&
>       |                                 ^~~~~~~~~~~~~
> src/esphome/components/deep_sleep/deep_sleep_component.h: In instantiation of 'void esphome::deep_sleep::EnterDeepSleepAction<Ts>::set_sleep_duration(V) [with V = setup()::<lambda()>; Ts = {}]':
> /config/esphome/fall-detector.yaml:141:57:   required from here
>   141 |                       sleep_duration: ${dozetime}
>       |                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~       ^
>   142 | 
>       |                                                          
>   143 |   on_shutdown:
>       |   ~~                                                     
> src/esphome/core/automation.h:108:64: warning: 'esphome::TemplatableFn<T, X>::TemplatableFn(F) requires !(convertible_to<F, T (*)('type_pack_expansion' not supported by simple_type_specifier'type_pack_expansion' not supported by direct_abstract_declarator)>) && (invocable<F, X ...>) && (convertible_to<typename std::invoke_result<F, X ...>::type, T>) && (is_empty_v<F>) && (default_initializable<F>) [with F = setup()::<lambda()>; T = long unsigned int; X = {}]' is deprecated: Lambda return type does not match TemplatableFn<T> — use the correct type in codegen [-Wdeprecated-declarations]
>   108 |   template<typename V> void set_##name(V name) { this->name##_ = name; }
> src/esphome/core/automation.h:110:39: note: in expansion of macro 'TEMPLATABLE_VALUE_'
>   110 | #define TEMPLATABLE_VALUE(type, name) TEMPLATABLE_VALUE_(type, name)
>       |                                       ^~~~~~~~~~~~~~~~~~
> src/esphome/components/deep_sleep/deep_sleep_component.h:167:3: note: in expansion of macro 'TEMPLATABLE_VALUE'
>   167 |   TEMPLATABLE_VALUE(uint32_t, sleep_duration);
>       |   ^~~~~~~~~~~~~~~~~
> src/esphome/core/automation.h:54:33: note: declared here
>    54 |                    "codegen")]] TemplatableFn(F) requires(!std::convertible_to<F, T (*)(X...)>) &&
>       |                                 ^~~~~~~~~~~~~
> 

### Which version of ESPHome has the issue?

2026.4.2

### What type of installation are you using?

Home Assistant Add-on

### What platform are you using?

ESP32

### Component causing the issue

esphome builder

### YAML Config

```yaml
substitutions:
  dozetime: "60min"
  criticalsleep: "24h"

.....

      - if: 
          condition: { lambda: 'return esp_sleep_get_wakeup_cause() != 7;' }
          then:
            - if:
                condition: 
                  switch.is_off: service_mode_switch
                then:            
                  - lambda: 'id(sleeping_state) = true;'
                  - deep_sleep.enter: 
                      id: sleep_mode
                      sleep_duration: ${dozetime}
```

### Anything in the logs that might be useful for us?

```txt
In file included from src/esphome/components/sensor/filter.h:9,
from src/esphome/components/sensor/sensor.h:8,
from src/esphome/components/adc/adc_sensor.h:3,
from src/esphome.h:3,
from src/main.cpp:3:
src/esphome/components/deep_sleep/deep_sleep_component.h: In instantiation of 'void esphome::deep_sleep::EnterDeepSleepAction::set_sleep_duration(V) [with V = setup()::<lambda()>; Ts = {}]':
/config/esphome/fall-detector.yaml:967:59: required from here
967 | sleep_duration: ${criticalsleep}
| ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^
968 | else:
| ~~~~~
969 | - if:
|
src/esphome/core/automation.h:108:64: warning: 'esphome::TemplatableFn<T, X>::TemplatableFn(F) requires !(convertible_to<F, T ()('type_pack_expansion' not supported by simple_type_specifier'type_pack_expansion' not supported by direct_abstract_declarator)>) && (invocable<F, X ...>) && (convertible_to<typename std::invoke_result<F, X ...>::type, T>) && (is_empty_v) && (default_initializable) [with F = setup()::<lambda()>; T = long unsigned int; X = {}]' is deprecated: Lambda return type does not match TemplatableFn — use the correct type in codegen [-Wdeprecated-declarations]
108 | template void set_##name(V name) { this->name##_ = name; }
src/esphome/core/automation.h:110:39: note: in expansion of macro 'TEMPLATABLE_VALUE_'
110 | #define TEMPLATABLE_VALUE(type, name) TEMPLATABLE_VALUE_(type, name)
| ^~~~~~~~~~~~~~~~~~
src/esphome/components/deep_sleep/deep_sleep_component.h:167:3: note: in expansion of macro 'TEMPLATABLE_VALUE'
167 | TEMPLATABLE_VALUE(uint32_t, sleep_duration);
| ^~~~~~~~~~~~~~~~~
src/esphome/core/automation.h:54:33: note: declared here
54 | "codegen")]] TemplatableFn(F) requires(!std::convertible_to<F, T ()(X...)>) &&
| ^~~~~~~~~~~~~
src/esphome/components/deep_sleep/deep_sleep_component.h: In instantiation of 'void esphome::deep_sleep::EnterDeepSleepAction::set_sleep_duration(V) [with V = setup()::<lambda()>; Ts = {}]':
/config/esphome/fall-detector.yaml:141:57: required from here
141 | sleep_duration: ${dozetime}
| ~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^
142 |
|
143 | on_shutdown:
| ~~
src/esphome/core/automation.h:108:64: warning: 'esphome::TemplatableFn<T, X>::TemplatableFn(F) requires !(convertible_to<F, T ()('type_pack_expansion' not supported by simple_type_specifier'type_pack_expansion' not supported by direct_abstract_declarator)>) && (invocable<F, X ...>) && (convertible_to<typename std::invoke_result<F, X ...>::type, T>) && (is_empty_v) && (default_initializable) [with F = setup()::<lambda()>; T = long unsigned int; X = {}]' is deprecated: Lambda return type does not match TemplatableFn — use the correct type in codegen [-Wdeprecated-declarations]
108 | template void set_##name(V name) { this->name##_ = name; }
src/esphome/core/automation.h:110:39: note: in expansion of macro 'TEMPLATABLE_VALUE_'
110 | #define TEMPLATABLE_VALUE(type, name) TEMPLATABLE_VALUE_(type, name)
| ^~~~~~~~~~~~~~~~~~
src/esphome/components/deep_sleep/deep_sleep_component.h:167:3: note: in expansion of macro 'TEMPLATABLE_VALUE'
167 | TEMPLATABLE_VALUE(uint32_t, sleep_duration);
| ^~~~~~~~~~~~~~~~~
src/esphome/core/automation.h:54:33: note: declared here
54 | "codegen")]] TemplatableFn(F) requires(!std::convertible_to<F, T ()(X...)>) &&
| ^~~~~~~~~~~~~
```

### Additional information

_No response_

--- a/esphome/components/deep_sleep/__init__.py
+++ b/esphome/components/deep_sleep/__init__.py
@@ -417,7 +417,7 @@ async def deep_sleep_enter_to_code(config, action_id, template_arg, args):
     paren = await cg.get_variable(config[CONF_ID])
     var = cg.new_Pvariable(action_id, template_arg, paren)
     if CONF_SLEEP_DURATION in config:
-        template_ = await cg.templatable(config[CONF_SLEEP_DURATION], args, cg.int32)
+        template_ = await cg.templatable(config[CONF_SLEEP_DURATION], args, cg.uint32)
         cg.add(var.set_sleep_duration(template_))
 
     if CONF_UNTIL in config:

--- a/tests/components/deep_sleep/common.yaml
+++ b/tests/components/deep_sleep/common.yaml
@@ -4,3 +4,9 @@ esphome:
       - deep_sleep.prevent
       - delay: 1s
       - deep_sleep.allow
+      - if:
+          condition:
+            lambda: 'return false;'
+          then:
+            - deep_sleep.enter:
+                sleep_duration: 60min

